### PR TITLE
Migrate to PHPUnit >9.3 and add support for PHP 8

### LINKED ISSUE #4: PHP 8.0 support
### Feature Request

|    Q        |   A
|------------ | ------
| New Feature | yes

#### Summary

To be prepared for the december release of PHP 8.0, this repository has some additional TODOs to be tested against the new major version.

In order to make this repository compatible, one has to follow these steps:

- [x] Modify `composer.json` to provide support for PHP 8.0 by adding the constraint `~8.0.0`
- [x] Modify `composer.json` to drop support for PHP less than 7.3
- [x] Modify `composer.json` to implement phpunit 9.3 which supports PHP 7.3+
- [x] Modify `.travis.yml` to ignore platform requirements when installing composer dependencies (simply add `--ignore-platform-reqs` to `COMPOSER_ARGS` env variable)
- [x] Modify `.travis.yml` to add PHP 8.0 to the matrix (*NOTE*: Do not allow failures as PHP 8.0 has a feature freeze since 2020-08-04!)
- [x] Modify source code in case there are incompatibilities with PHP 8.0

### PR DESCRIPTION
**Pull request by internalsystemerror**
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | no
| RFC           | no
| QA            | yes

### Description

This PR upgrades the testsuite to support PHPUnit 9.3. Laminas Coding Standards v2 is now used and support for PHP 8 has been tested and added.

Note that there is already an existing PR for PHP8 #5 however this has sat out of date for some time and does not include the recent GHA changes that were included. This PR includes those, and above that refactors the unit tests to reflect the newer direction that PHPUnit is going.

### CODE REVIEW

File: src/TwigEnvironmentFactory.php [hunk 1]
```diff
@@ -245,11 +239,13 @@ private function loadRuntimeLoader($runtimeLoader, ContainerInterface $container
...
-                RuntimeLoaderInterface::class,
-                is_object($runtimeLoader) ? get_class($runtimeLoader) : gettype($runtimeLoader)
-            ));
+            throw new Exception\InvalidRuntimeLoaderException(
+                sprintf(
+                    'Twig runtime loader must be an instance of %s; "%s" given,',
+                    RuntimeLoaderInterface::class,
+                    is_object($runtimeLoader) ? get_class($runtimeLoader) : gettype($runtimeLoader)
+                )
+            );
```
Comment (Xerkus) [inline src/TwigEnvironmentFactory.php]: Is this change dictated by our coding standard? Original formatting should be just fine, I think.

Comment (internalsystemerror) [reply src/TwigEnvironmentFactory.php]: I just ran cs-fix across the project, either that or PhpStorm did it, though I typically leave settings as default for mezzio/laminas repos.

File: src/TwigExtension.php [hunk 2]
```diff
@@ -81,78 +73,64 @@ public function getFunctions() : array
...
+     * @param array $routeParams
+     * @param array $queryParams
+     * @param array $options Can have the following keys:
      *                             - reuse_result_params (bool): indicates if the current
      *                             RouteResult parameters will be used, defaults to true
-     *
-     * @return string
      */
-    public function renderUri(
+    public function renderUrl(
```
Comment (Xerkus) [inline src/TwigExtension.php]: Method ordering change is not needed for this PR and it confuses diff with unnecessary changes, please revert

File: src/TwigRenderer.php [hunk 3]
```diff
@@ -112,14 +97,27 @@ public function render(string $name, $params = []) : string
...
     /**
-     * Add a path for template
+     * Normalize namespaced template.
      *
-     * @param string      $path
-     * @param string|null $namespace
+     * Normalizes templates in the format "namespace::template" to
+     * "@namespace/template".
+     */
+    public function normalizeTemplate(string $template): string
```
Comment (Xerkus) [inline src/TwigRenderer.php]: Again, unnecessary method order change, please revert.

File: test/ConfigProviderTest.php [hunk 4]
```diff
@@ -15,17 +15,13 @@
...
-    /**
-     * @var ConfigProvider
-     */
+    /** @var ConfigProvider */
     private $provider;
 
-    protected function setUp() : void
-    {
-        $this->provider = new ConfigProvider();
-    }
```
Comment (Xerkus) [inline test/ConfigProviderTest.php]: Our tests have `setUp` as the first method in tests across the org, please revert the change and bring them back up for consistency here and in the rest of the PR.

File: test/TestAsset/test/test.js [hunk 5]
```diff
@@ -1,3 +1,5 @@
 {
-  "test": "twig"
+    'test';
+:
+    'twig';
 }
```
Comment (Xerkus) [inline test/TestAsset/test/test.js]: this looks wrong

File: test/TwigEnvironmentFactoryTest.php [hunk 6]
```diff
@@ -81,52 +77,83 @@ public function testUsesDebugConfiguration()
...
+            false,
+            true,
+            true,
+            true,
+            true,
+            true,
+            false,
+            true,
+            true,
+        );
```
Comment (Xerkus) [inline test/TwigEnvironmentFactoryTest.php]: Call order does not matter. This test will break if order of calls changes, which makes test brittle as compared to prophecy mock it replaces.

Could you refactor this and other similar changes to use a map or callback instead?

File: src/TwigEnvironmentFactory.php (hunk 1 above)
Comment (internalsystemerror) [reply src/TwigEnvironmentFactory.php]: fix incoming

File: test/TwigEnvironmentFactoryTest.php [hunk 7]
```diff
@@ -191,83 +215,87 @@ public function invalidConfiguration()
...
     }
 
-    public function testRaisesExceptionForInvalidTimezone()
+    /**
+     * @throws LoaderError
+     */
+    public function testRaisesExceptionForInvalidTimezone(): void
     {
-        $tz = 'Luna/Copernicus_Crater';
+        $tz     = 'Luna/Copernicus_Crater';
```
Comment (Xerkus) [inline test/TwigEnvironmentFactoryTest.php:220]: Oh no. This made me realize we will have to deal with timezones on other planetary bodies in the next decade or two.

**Review by Xerkus — CHANGES_REQUESTED** [review]
Looks fine except for few minor points:
- change of methods order is unnecessary and really throws diff off. I would prefer to avoid those unnecessary changes.
- ordering of calls for `has` or even `get` methods is unimportant and it makes tests brittle for no gain with the switch to mock objects. map or callback would be preferable.
- I checked across number of components and `setUp()` and `tearDown()` are consistently at the top of the test cases, similar to constructors. I request to bring them back up for consistency.

File: src/TwigExtension.php (hunk 2 above)
Comment (internalsystemerror) [reply src/TwigExtension.php]: This looks like I had format on save enabled in my IDE. Thanks, I'll resolve.

Comment (internalsystemerror) [reply src/TwigExtension.php]: corrected

File: src/TwigRenderer.php (hunk 3 above)
Comment (internalsystemerror) [reply src/TwigRenderer.php]: corrected

File: test/ConfigProviderTest.php (hunk 4 above)
Comment (internalsystemerror) [reply test/ConfigProviderTest.php]: corrected

File: test/TestAsset/test/test.js (hunk 5 above)
Comment (internalsystemerror) [reply test/TestAsset/test/test.js]: this should have been json not js, hence why it was invalidly formatted. corrected to the original syntax, and changed the file ext to json

File: test/TwigEnvironmentFactoryTest.php (hunk 7 above)
Comment (internalsystemerror) [reply test/TwigEnvironmentFactoryTest.php:220]: woah

File: test/TwigEnvironmentFactoryTest.php (hunk 6 above)
Comment (internalsystemerror) [reply test/TwigEnvironmentFactoryTest.php]: Yes that's a much better idea, will do.

File: test/TestAsset/test/test.js (hunk 5 above)
Comment (Xerkus) [reply test/TestAsset/test/test.js]: it was looking for `.js` tho:
```
Twig\Error\LoaderError: Unable to find template "@test/test.js" (looked into: /github/workspace/test/TestAsset/test).
```

File: test/TwigEnvironmentFactoryTest.php (hunk 6 above)
Comment (internalsystemerror) [reply test/TwigEnvironmentFactoryTest.php]: Refactored all uses of the new mocks to make the tests less brittle as suggested.

**Comment by internalsystemerror** [comment]
Thanks for reviewing this PR, I've tried to address all your concerns and have implemented your suggestions. This should make the diff easier to digest. Let me know if there's anything else, I'm happy to keep going as needed.

File: src/TwigExtensionFactory.php [hunk 8]
```diff
@@ -19,36 +19,32 @@
...
-        ) {
-            throw new InvalidConfigException(sprintf(
-                'Missing required `%s` dependency.',
-                ServerUrlHelper::class
-            ));
+        $serverUrlHelper = $container->has(ServerUrlHelper::class) ? ServerUrlHelper::class : (
+        $container->has(\Zend\Expressive\Helper\ServerUrlHelper::class)
+            ? \Zend\Expressive\Helper\ServerUrlHelper::class
+            : null
+        );
```
Comment (Xerkus) [inline src/TwigExtensionFactory.php]: This bit is hard to read, I would suggest to reformat it like this and use alias:
```suggestion
        $serverUrlHelper = $container->has(ServerUrlHelper::class) 
            ? ServerUrlHelper::class
            : ($container->has(ZfServerUrlHelper::class) ? ZfServerUrlHelper::class : null);
```
 
Suggesting `Zf` instead of `Zend` because it would otherwise form `ZendServer` which is unrelated product

File: src/TwigExtensionFactory.php [hunk 9]
```diff
@@ -19,36 +19,32 @@
...
-        ) {
-            throw new InvalidConfigException(sprintf(
-                'Missing required `%s` dependency.',
-                UrlHelper::class
-            ));
+        $urlHelper = $container->has(UrlHelper::class) ? UrlHelper::class : (
+        $container->has(\Zend\Expressive\Helper\UrlHelper::class)
+            ? \Zend\Expressive\Helper\UrlHelper::class
+            : null
+        );
```
Comment (Xerkus) [inline src/TwigExtensionFactory.php]:
```suggestion
        $urlHelper = $container->has(UrlHelper::class)
            ? UrlHelper::class
            : ($container->has(ZfUrlHelper::class) ? ZfUrlHelper::class : null);
```

File: test/TwigEnvironmentFactoryTest.php [hunk 10]
```diff
@@ -19,59 +19,55 @@
...
-    protected function setUp() : void
+    protected function setUp(): void
     {
-        $this->container = $this->prophesize(ContainerInterface::class);
+        $this->container = $this->createMock(ContainerInterface::class);
     }
 
-    public function testCallingFactoryWithNoConfigReturnsTwigEnvironmentInstance()
+    /**
+     * @throws LoaderError
```
Comment (Xerkus) [inline test/TwigEnvironmentFactoryTest.php]: We don't annotate throws on tests. It is not useful, just some extra noise

File: test/TwigEnvironmentFactoryTest.php [hunk 11]
```diff
@@ -19,59 +19,55 @@
...
         $config = ['debug' => true];
-        $this->container->has('config')->willReturn(true);
-        $this->container->get('config')->willReturn($config);
-        $this->container->has(TwigExtension::class)->willReturn(false);
-        $this->container->has(\Zend\Expressive\Twig\TwigExtension::class)->willReturn(false);
-        $this->container->has(ServerUrlHelper::class)->willReturn(false);
-        $this->container->has(\Zend\Expressive\Helper\ServerUrlHelper::class)->willReturn(false);
-        $this->container->has(UrlHelper::class)->willReturn(false);
-        $this->container->has(\Zend\Expressive\Helper\UrlHelper::class)->willReturn(false);
+        $this->container->expects(self::atLeastOnce())->method('has')->withAnyParameters()->willReturnMap(
```
Comment (Xerkus) [inline test/TwigEnvironmentFactoryTest.php]: `withAnyParameters()` is default I believe, so it can be dropped here and in other similar uses
```suggestion
        $this->container->expects(self::atLeastOnce())->method('has')->willReturnMap(
```

**Review by Xerkus — CHANGES_REQUESTED** [review]

File: test/TwigEnvironmentFactoryTest.php (hunk 10 above)
Comment (internalsystemerror) [reply test/TwigEnvironmentFactoryTest.php]: removed all @throws additions in tests

File: test/TwigEnvironmentFactoryTest.php (hunk 11 above)
Comment (internalsystemerror) [reply test/TwigEnvironmentFactoryTest.php]: done for all

**Comment by internalsystemerror** [comment]
I've implemented all points. Let me know for any more, otherwise do you need me to rebase and squash to a single commit?

**Review by Xerkus — APPROVED** [review]
